REF/MAINT: deprecate sjoin op param in favor of predicate

adriangb opened the pull request:
Closes #1525

brendan-ward reviewed (changes requested):
@adriangb thanks for working on this, I think `predicate` is much more clear now in this context than `op`.  👍 

Since this deprecates a parameter in a top-level API, it might be good to target this to 0.9 instead of next 0.8.x release (only effect here being that we wait a bit to merge).

Will also need slight update in upcoming `sjoin` documentation @martinfleis

adriangb commented:
Where are we with this @martinfleis ? Do you want me to resolve conflicts and fix anything else?

martinfleis reviewed (commented):
Ah, this has slipped from my mind. Should've been in 0.9.0.

Can you just change `op` to `predicate` in the user guide? https://geopandas.readthedocs.io/en/latest/docs/user_guide/mergingdata.html#spatial-joins

adriangb commented on geopandas/tools/sjoin.py:
```diff
@@ -21,19 +27,32 @@ def sjoin(
+        `predicate`.
     """
+    if "op" in kwargs:
+        warnings.warn(
```
Swapped

adriangb commented:
Hey @martinfleis, I made the doc edits and swapped the `UserWarning -> FutureWarning`.

I also realized that there's 2 edge cases that we weren't checking for:
- Both `op` and `predicate` are specified. I added a `UserWarning` for this that is only emitted if `predicate!="intersects"` (i.e. not the default) _and_ `op!=predicate`. This also includes the deprecation message. Let me know if you'd rather this be a `FutureWarning`.
- Extra `**kwargs` are specified. This will now raise a `ValueError` that lists out the extra kwargs.

martinfleis reviewed (approved):
Thanks @adriangb!

adriangb commented:
@brendan-ward bumping again, hoping this doesn't get left out of the next release

martinfleis commented:
@adriangb can you make sure this is mergeable (no conflicts, green CI)?

adriangb commented:
yeah let me resolve conflicts

adriangb commented:
Conflicts resolved. CI failures look like the same Matplotlib issue as #2066

brendan-ward commented on geopandas/tools/sjoin.py:
```diff
@@ -84,6 +94,33 @@ def sjoin(
+            warnings.warn(override_message, UserWarning, stacklevel=4)
+        else:
+            warnings.warn(deprectation_message, FutureWarning, stacklevel=4)
+    if kwargs:
```
Would be good to model this off other functions, e.g.,  

```
>>> df.contains(df.geometry, foo='bar')
...
TypeError: contains() got an unexpected keyword argument 'foo'
```

adriangb replied:
makes sense, simpler as well

brendan-ward reviewed (commented):
@adriangb thanks for working on this!

A few minor suggestions, but then otherwise this looks good to me.  The CI failures seem unrelated.

adriangb commented on geopandas/tools/sjoin.py:
```diff
@@ -24,15 +30,19 @@ def sjoin(
     rsuffix : string, default 'right'
         Suffix to apply to overlapping column names (right GeoDataFrame).
+    **kwargs: dict
+        For handling of deprecated `op` param, which will be replaced by
```
ok

adriangb commented on geopandas/tools/sjoin.py:
```diff
@@ -84,6 +94,33 @@ def sjoin(
+            )
+            warnings.warn(override_message, UserWarning, stacklevel=4)
+        else:
+            warnings.warn(deprectation_message, FutureWarning, stacklevel=4)
```
great catch!

adriangb replied:
I added:

```python
@pytest.mark.parametrize("predicate_kw", ["predicate", "op"])
```

To the `test_inner` test. I think this is the easiest way to catch this bug.

adriangb commented:
All tests are passing now w/ master merged in, and I believe I've addressed all of the feedback. Please ping me if there's anything else needed @brendan-ward @martinfleis

adriangb commented:
@martinfleis is there anything blocking a merge?

martinfleis commented:
> @martinfleis is there anything blocking a merge?

I'd like to have another review approval here.

adriangb commented:
Got it. Happy to wait 😄

martinfleis commented:
Here we go :).

Thanks @adriangb!
